python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py

```diff
@@ -0,0 +1,350 @@
+        return MaterializeResult(metadata={"foo": "bar"})
+
+    assert materialize(
+        [asset_with_type_annotation], resources={"io_manager": TestingIOManager()}
+    ).success
+
+    @multi_asset(outs={"one": AssetOut(), "two": AssetOut()})
+    def multi_asset_with_outs_and_type_annotation() -> Tuple[MaterializeResult, MaterializeResult]:
+        return MaterializeResult(asset_key="one"), MaterializeResult(asset_key="two")
+
+    assert materialize(
+        [multi_asset_with_outs_and_type_annotation], resources={"io_manager": TestingIOManager()}
+    ).success
+
+    @multi_asset(specs=[AssetSpec("one"), AssetSpec("two")])
+    def multi_asset_with_specs_and_type_annotation() -> Tuple[MaterializeResult, MaterializeResult]:
+        return MaterializeResult(asset_key="one"), MaterializeResult(asset_key="two")
+
+    assert materialize(
+        [multi_asset_with_specs_and_type_annotation], resources={"io_manager": TestingIOManager()}
+    ).success
+
+    @multi_asset(specs=[AssetSpec("one"), AssetSpec("two")])
+    def multi_asset_with_specs_and_no_type_annotation():
+        return MaterializeResult(asset_key="one"), MaterializeResult(asset_key="two")
+
+    assert materialize(
+        [multi_asset_with_specs_and_no_type_annotation],
+        resources={"io_manager": TestingIOManager()},
+    ).success
+
+
+@pytest.mark.skip("generator return types are interpreted as Any")
+def test_generator_return_type_annotation():
+    class TestingIOManager(IOManager):
+        def handle_output(self, context, obj):
+            assert context.dagster_type.is_nothing
+            return None
+
+        def load_input(self, context):
+            return 1
+
+    @asset
+    def generator_asset() -> Generator[MaterializeResult, None, None]:
+        yield MaterializeResult(metadata={"foo": "bar"})
+
+    materialize([generator_asset], resources={"io_manager": TestingIOManager()})
+
+
+def test_direct_invocation_materialize_result():
+    @asset
+    def my_asset() -> MaterializeResult:
+        return MaterializeResult(metadata={"foo": "bar"})
+
+    res = my_asset()
+    assert res.metadata["foo"] == "bar"
+
+    @multi_asset(specs=[AssetSpec("one"), AssetSpec("two")])
+    def specs_multi_asset():
+        return MaterializeResult(asset_key="one", metadata={"foo": "bar"}), MaterializeResult(
+            asset_key="two", metadata={"baz": "qux"}
+        )
+
+    res = specs_multi_asset()
+    assert res[0].metadata["foo"] == "bar"
+    assert res[1].metadata["baz"] == "qux"
+
+    @multi_asset(outs={"one": AssetOut(), "two": AssetOut()})
+    def outs_multi_asset():
+        return MaterializeResult(asset_key="one", metadata={"foo": "bar"}), MaterializeResult(
+            asset_key="two", metadata={"baz": "qux"}
+        )
+
+    res = outs_multi_asset()
+    assert res[0].metadata["foo"] == "bar"
+    assert res[1].metadata["baz"] == "qux"
+
+
+@pytest.mark.skip("direct invocation for generators does not work yet")
```

Comment:
marked as skipped because direct invocation of yielded MaterializeResult doesn't work. See PR description for more info

Reply:
to add on top of the description, this is where it errors: https://github.com/dagster-io/dagster/blob/9ec0f16d9c83ed3d75ee609ba222e30e22ba7170/python_modules/dagster/dagster/_core/definitions/op_invocation.py#L424

either
1) we handle `MaterializeResult` (and `AssetCheckResult` as well i believe) specially, e.g. to let the output type checking block ([here](https://github.com/dagster-io/dagster/blob/9ec0f16d9c83ed3d75ee609ba222e30e22ba7170/python_modules/dagster/dagster/_core/definitions/op_invocation.py#L414-L427)) know that we should skip that output check, such as record it somehow in the "outputs_seen". but that doesn't feel great to me.
or 
2) we handle the dagster type resolution for generator correctly. [here](https://github.com/dagster-io/dagster/blob/9ec0f16d9c83ed3d75ee609ba222e30e22ba7170/python_modules/dagster/dagster/_core/definitions/inference.py#L65), all generator return type annotation are being inferred as "Any", so the fix in https://github.com/dagster-io/dagster/pull/16883 don't affect the dagster_type when the return type hint is a generator.